### ISSUE TITLE
[Meta] PHP 7.3 support

### ISSUE THREAD

**Issue by julienfalque** [FriendsOfPHP/PHP-CS-Fixer#3697]
This issue lists PHP 7.3 changes that are relevant to this project and tracks related issues and PR.

#### Syntax changes that require new tests to make sure the tool is compatible with PHP 7.3:

- [x] [Flexible Heredoc and Nowdoc Syntaxes](https://wiki.php.net/rfc/flexible_heredoc_nowdoc_syntaxes) (#3915)
- [x] [list() Reference Assignment](https://wiki.php.net/rfc/list_reference_assignment)
- [x] `instanceof` now allows literals as the first operand, in which case the result is always `false`.
- [x] [Allow a trailing comma in function calls](https://wiki.php.net/rfc/trailing-comma-function-calls)

Progress tracker: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4184

#### Relevant changes that could lead to new features (not required to make the tool compatible with PHP 7.3):

- [ ] `hrtime` function has been added
- [ ] `is_countable` function has been added => #4236
- [ ] add trailing comma to multiline function calls => #4237

#### Sources:
- https://github.com/php/php-src/blob/PHP-7.3/NEWS
- https://github.com/php/php-src/blob/PHP-7.3/UPGRADING

**Comment by ntzm** [comment]
Would be nice to get a fixer to pass the new `JSON_THROW_ON_ERROR` flag to `json_` methods when 7.3 is released

**Comment by julienfalque** [comment]
Could be nice indeed, but that should be a separate feature request. This issue mostly tracks changes making _existing_ code compatible with PHP 7.3 :)

**Comment by keradus** [comment]
and then, clean up manual handling of error, huh? gonna be tricky ;(

**Comment by SpacePossum** [comment]
- `is_countable` function
- [Allow a trailing comma in function calls](https://wiki.php.net/rfc/trailing-comma-function-calls)

The last one will be a bit of work here I think.
It also seems to be missing from the releases note, maybe that gets corrected before the release though.

**Comment by julienfalque** [comment]
Thanks @SpacePossum, I edited the list. Note that I think we don't need to care about `is_countable()` to be compatible with PHP 7.3. Having rules for it would be a separate feature request, and one was just submitted by @ntzm: #3711 ;)

**Comment by keradus** [comment]
whenever we care about native functions, we need to take care of `is_countable` now as well.

**Comment by SpacePossum** [comment]
I wonder if we could help people mediate issues like these https://externals.io/message/102147 when going from PHP7.2 to 7.3, tricky part is to identify when the fix needs to be applied and not re-apply it on a second run on PHP7.3.....

**Comment by ntzm** [comment]
`net_get_interfaces` is added in 7.3 as well: https://github.com/php/php-src/blob/master/NEWS

**Comment by ntzm** [comment]
Alpha 1 is out, spotted a bug in `token_get_all` with the `TOKEN_PARSE` flag that was breaking a lot of php-cs-fixer, reported [here](https://bugs.php.net/bug.php?id=76437)

**Comment by gharlan** [comment]
> [Flexible Heredoc and Nowdoc Syntaxes](https://wiki.php.net/rfc/flexible_heredoc_nowdoc_syntaxes)

I want to create a fixer for heredoc/nowdoc indentation.

Input:

```php
function foo() {
    return [
        'foo',
        <<<'EOF'
abcdef
    ghijkl
mnopqr
EOF
        ,
        'bar',
        'baz',
    ];
}
```

Which indentation would you prefer?

v1:

```php
function foo() {
    return [
        'foo',
        <<<'EOF'
        abcdef
            ghijkl
        mnopqr
        EOF
        ,
        'bar',
        'baz',
    ];
}
```

v2:

```php
function foo() {
    return [
        'foo',
        <<<'EOF'
            abcdef
                ghijkl
            mnopqr
            EOF
        ,
        'bar',
        'baz',
    ];
}
```

And should the same fixer replace the comma (`EOF,`), or should it be done by another (new) fixer, or by an existing fixer about spaces around `,`?

**Comment by ntzm** [comment]
I would personally prefer V1, and I think the comma change should be done on fixer `no_whitespace_before_comma_in_array` on 7.3+ of course

**Comment by julienfalque** [comment]
I'd prefer v2. Indentation should probably be configurable. Handling of the comma should be done by `no_whitespace_before_comma_in_array` indeed.

**Comment by ntzm** [comment]
`hrtime` function has also been added

**Comment by keradus** [comment]
nightly job is green now

**Comment by Slamdunk** [comment]
> Sources:
> 
>     https://github.com/php/php-src/blob/master/NEWS
>     https://github.com/php/php-src/blob/master/UPGRADING

Correct links are:

* https://github.com/php/php-src/blob/PHP-7.3/NEWS
* https://github.com/php/php-src/blob/PHP-7.3/UPGRADING

**Comment by DeepDiver1975** [comment]
Facing this issue: already addressed or reported?

```
eprecated: strrchr(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior in /home/travis/build/sabre-io/dav/vendor/friendsofphp/php-cs-fixer/Symfony/CS/Utils.php on line 115

**Comment by ntzm** [comment]
@DeepDiver1975 You are using PHP-CS-Fixer 1.x, please update to 2.x by reading the guide here: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.12/UPGRADE.md#upgrade-guide-from-1x-to-20

**Comment by TomasVotruba** [comment]
Just FIY, blocking PHP 7.3 makes fail builds that want to use it any way:
https://travis-ci.org/phpmyadmin/phpmyadmin/jobs/460537777

**Comment by keradus** [comment]
well, we gave you very same hint when you asked month ago ;)
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4051#issuecomment-430970261

**Comment by winkbrace** [comment]
Isn't there a way to force run cs fixer on php 7.3? Right now I have to downgrade php version or remove  the library. :(
The problem I'm facing is that cs fixer is part of the pre-commit hooks of one project where the other developers are still on php 7.2 anyway. But on another project I require php 7.3. I could use virtual environments of course or docker images, but a simple flag to force run it would be much simpler.

**Comment by jakzal** [comment]
Are all the items from the above list actually necessary to implement in order to get php-cs-fixer working on PHP 7.3? The build is passing after all. Suggestion to ignore platform requirements also hints that php-cs-fixer actually works on PHP 7.3. 

Forgive me if I've missed anything, but why not add 7.3 to supported versions?

**Comment by theofidry** [comment]
@JLHwung?

**Comment by julienfalque** [comment]
@jakzal @theofidry Please see https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3323#issuecomment-355007563

**Comment by julienfalque** [comment]
Indeed, this issue is about PHP 7.3 support, that is:
1. making sure all existing fixers are compatible with new PHP 7.3 syntaxes;
2. list new features we can implement regarding those new syntaxes.

So yes, only 1 is required to allow installing the tool on PHP 7.3 :)

**Comment by dmvdbrugge** [comment]
@julienfalque maybe update the initial post to show the separation? (Just for clarification.)

**Comment by andrerom** [comment]
Or split in two issues to prioritize to unblock (clean) php 7.3 usage?

**Comment by JLHwung** [comment]
@theofidry the thumbdown is for:

- winkbrace has gave us a workaround [here](https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3697#issuecomment-446142450), I see no reason it is actually "blocking 7.3 support".
- Test is not breaking in PHP 7.3 does not means that the support for PHP 7.3 is ready. As you can see there is some new feature of PHP 7.3 that cs-fixer does not support yet. I am glad the the developer are very cautious about the platform requirement.

As an OSS the php-cs-fixer owns us nothing to any prompt support for any new version of PHP. Don't pressure on maintainers and if you can and you would, raise a PR to contribute the project.

**Comment by theofidry** [comment]
@JLHwung thanks for clarifying, as you can see our beef was more against new fixers/rules which are unnecessary for having a 7.3 support.

>Don't pressure on maintainers and if you can and you would, raise a PR to contribute the project.

Sure, but it's better to be clear about what is needed first right :)

**Comment by jakzal** [comment]
> winkbrace has gave us a workaround here, I see no reason it is actually "blocking 7.3 support"

The workaround won't work for me unfortunately. As in my case php-cs-fixer is installed as a composer dependency and I don't want to ignore platform reps for other packages. As much as I'd love to install php-cs-fixer as a phar, it's not happening on this project.

So yeah, it is blocking my move to PHP 7.3, but please don't treat it as putting a pressure on maintainers. I asked a genuine question as I don't understand why a workaround which is equivalent to allowing 7.3 in composer.json is better than actually allowing php 7.3.

**Comment by julienfalque** [comment]
@jakzal Maybe an alternative solution could be that you install the tool with a dedicated `composer.json` in another directory (e.g. `tools/php-cs-fixer`). Another benefit from doing this is that PHP CS Fixer's dependencies won't interfere with your project's dependencies anymore.

**Comment by derrabus** [comment]
> Suggestion to ignore platform requirements also hints that php-cs-fixer actually works on PHP 7.3.

It probably boils down to this: You can safely use CS Fixer on PHP 7.3 as long as the code that you're fixing is not using any language features of PHP 7.3 yet.

**Comment by localheinz** [comment]
The easiest workarounds are

- remove `friendsofphp/php-cs-fixer` on builds running on PHP 7.3
- require `friendsofphp/php-cs-fixer` on a build not running PHP 7.3

There is only a need to run code style checks and fixes on a single build, not all of them.

**Comment by andrerom** [comment]
@localheinz This is not just about builds, it's also about being able to to install applications in the first place _(mainly talking about composer create-project as way to instal apps, but also git clone + composer install, in cases you need dev packages for dev usage, where we *really* want to error out if requirements are not meet)_

> Allow the installation but show an error when running on a version with not fully supported syntax stating that CS Fixer will only work if you don't use those new features and maybe urging the user to send a PR.

This is definitely the right way to go with this.

**Comment by Nothing-Works** [comment]
@winkbrace can I set `PHP_CS_FIXER_IGNORE_ENV=1` as default somewhere? so I can use it with vscode extension?

**Comment by winkbrace** [comment]
@Nothing-Works It's en environment variable that must be set. I'm not familiar with vscode, but in phpstorm I have to use php cs fixer as an external tool that I call with a prefix: `PHP_CS_FIXER_IGNORE_ENV=1; /Users/winkbrace/.composer/vendor/bin/php-cs-fixer`

**Comment by Zheaoli** [comment]
excuse me, is there any roadmap to support PHP 7.3?

**Comment by theofidry** [comment]
@Zheaoli the description of the issue

**Comment by kelunik** [comment]
> It probably boils down to this: You can safely use CS Fixer on PHP 7.3 as long as the code that you're fixing is not using any language features of PHP 7.3 yet.

This. Requirements should be runtime requirements and not state the code that can be fixed. If you write a fixer for Ruby code, you wouldn't write Ruby into the requirements, no?

**Comment by andrerom** [comment]
@teohhanhui No judgment given, but care to expand on why you vote down several comments which seems to make sense from package consumer perspective?

**Comment by devdrops** [comment]
Hi!

I've just stumbled into this scenario (installing `friendsofphp/php-cs-fixer` on PHP 7.3) and noticed this issue.

Based on the list [in the first message](https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3697#issue-314315352), is the list up-to-date? If so, I'd like to help by checking some of these items :wink:

**Comment by SpacePossum** [comment]
Hi!

I would say that my last [comment](https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3697#issuecomment-446555136) is the most up-to-date. 
Basically it boils down to:
- finding fixers/code that do not handle the new `Allow a trailing comma in function calls`  syntax well
- fix the issues

everything else from PHP 7.3 are nice-to-haves we can add later on :)

**Comment by keradus** [comment]
way too late probably, but just to clarify the policy.

> Also this issue comes out about CS Fixer every time a new PHP version is released.

> I agree with @andrerom that a warning would be better for most cases.

No. Warning is not enough. PHP CS Fixer must not break the code unless it was told otherwise. So, running the Fixer when there are high chances to break the codebase of user with just informing him "btw, I don't know what I'm doing, maybe I broke sth" is not acceptable. Especially for that, for two behaviors we have that could change/break the codebase, we have 2 special flags - `allow risky` and `ignore env` - that user of tool must explicitly allow the Tool to run.

Saying that, stability of the Tool is top prio. When user is not allowing the risky behaviour, we promise to not break the codebase and we do our best on that.
(btw, thanks @JLHwung for nice words)

I do know that topic of 7.3 is burning - but please, keep in mind that this tool is driven by community and made in free time of ppl. For that, resources are simply limited, and all the help is appreciated (not only raising the PRs, but also reviewing them, looking for missing edge cases, doing regression tests, ...).
Thankfully, there was some movement recently about adding support for new syntax, thus topic is not frozen.

**Comment by julienfalque** [comment]
@SpacePossum I see you marked the following tasks as completed:
* `list()` Reference Assignment
* `instanceof` now allows literals as the first operand, in which case the result is always `false`

I don't remember seeing related PRs for those topics, are you sure they are covered enough by tests?

**Comment by SpacePossum** [comment]
We already supported list reference assigned as I thought it was part of PHP7.2... ^_^
Some other edge I found when verifying this can be found here; https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4134 together with additional tests. Please let me know if I missed any cases.

The second one for `instanceof` I could not find any issues with the current fixers. But indeed no tests for it. I'll uncheck the item.

**Comment by CurtisBelt** [comment]
I use VSCode extension `junstyle.php-cs-fixer`, and was able to get up and running again by setting `export PHP_CS_FIXER_IGNORE_ENV=1` in my equivalent of a `.bash_profile`, and then restarting VSCode.

Maybe this will help others who are patiently waiting for the next release but are OK with the unsupported features :+1: 

P.S. Don't forget to remove the flag from your `.bash_profile` once 7.3 is supported.

**Comment by keradus** [comment]
Once more, just to stress that.
While it is possible to bypass all guards and run PHP CS Fixer anyway, fixing code written in 7.3 may lead to crash, eg
`PHP Parse error: syntax error, unexpected ','` after fixing. Gladly, PHP CS Fixer by default check if file is valid after the process, and if not - it doesn't save it.

**Comment by keradus** [comment]
I have created an integration test for 7.3: #4184
it shows how far we are to let our Fixers understand code written in 7.3 - after all the fixes we already made, there are almost 20 Fixers that still needs adjustment.

If anyone is willing to help, please visit #4184 and simply pick one of listed fixers, example code that is failing for given Fixer is provided in that PR as well.

If anyone is willing to ask why there is no support of PHP 7.3 and s/he wants to have it, please visit :point_up_2: paragraph ;)

**Comment by kelunik** [comment]
I don't care about PHP 7.3 syntax support (for now, as the current target is lower PHP versions), just about being able to install `php-cs-fixer` via Composer without the need for `--ignore-platform-reqs`. I'm fine with the runtime environment flag, but installation currently isn't possible without ignoring _all_ platform requirements. If I ignore all, I also won't get extension requirements. I think requiring the runtime flag is enough, there's no need for the additional installation prevention via `"require"` of a lower PHP version.

**Comment by kelunik** [comment]
@keradus You're entirely right, IMO all packages should only have a lower bound for PHP versions. That's also why [we use](https://github.com/amphp/amp/blob/216315ee5c59be6847af3391f11f04404d1ce659/composer.json#L36) only lower bounds for our packages, no upper bounds. PHP is pretty stable and unlike Composer dependencies, it's not installed as a dependency by Composer, but already installed on the system.

Besides that, the issue seems to be only PHP 7.3 _syntax_ that's not yet fully supported. Processing of PHP 7.2 and lower syntax seems to work fine, even on PHP 7.3, no?

**Comment by SpacePossum** [comment]
There are a lot different ideas that could be considered, like setting upper bound PHP support on fixer/rule level, etc. In the end however, all this takes up time that could be used spend on fixing the issues themselves.

Feel free to raise new concerns about what doesn't work when fixing PHP 7.3 code, but lets end the discussion about how to bypass the safety-mechanics preventing installing PHP-CS-Fixer on a PHP7.3 setup with composer (instructions are within the comments above), as no new arguments have been made for a while and no new insights have come from this to change the current ways.

Last days a lot of progress has been made towards PHP 7.3 support and I would really like to not loose momentum and focus because of the discussion here. We try to resolve the issues and prepare for a release with PHP 7.3 support so please help out with that :)

**Comment by localheinz** [comment]
@keradus 

🚀

**Comment by danger89** [comment]
@SpacePossum I agree, face-it, you can better have small iterations and have progress, instead of standing still too long. Let's have 7.3 support!
